init gnutls from ed keypair

tewinget opened the pull request:
Initialize GNUTLS creds from Ed keypair and confirm key allowed on incoming connection.

tewinget committed:
init gnutls from ed keypair

A GNUTLSCreds can now be initialized from a pair of Ed25519 seed/pubkey.

This enables using libquic in "raw public key" TLS mode, which is
useful/needed for Lokinet.

GNUTLS says it checks that the keys given match one another, but
it does not appear to do so.

This commit includes lokinet-specific ALPN code, as testing it in this
codebase is easier at the moment.  The ALPN specifics and Lokinet
specifics should be separated and the Lokinet specifics should move into
Lokinet.

GNUTLSCreds in raw pubkey mode accepts a verify callback which, given the
other side's public key and requested ALPN will determine whether the
connection should be allowed.

Use case for Lokinet:
  clients send the client ALPN during negotiation,
  relays send the relay ALPN on outbound, and
  relays set both for inbound.
  This is used to determine whether we are being connected to by
  a relay or a client, and check keys / setup lokinet accordingly.

if we need gnutls debug logs, we can set
GNUTLS_DEBUG_LEVEL=level (0-99, >10 is full verbosity it seems)
and gnutls will log to stderr

dr7ana commented on tests/008-ed-keys.cpp:
```diff
@@ -0,0 +1,147 @@
+
+            REQUIRE_NOTHROW(client_endpoint->connect(client_remote, client_tls));
+
+            REQUIRE(f.wait_for(20ms) == std::future_status::ready);
```
Are these sleeps being used to give the client time to establish a connection? You could use the same logic in the other tests like [001](https://github.com/oxen-io/oxen-libquic/blob/857188c5d21ea01819c0d88258b84f592ed05125/tests/001-handshake.cpp#L190), where the `client_tls_policy` is given a callback that signals "connection ready"

tewinget replied:
This is a future wait with a timeout, rather than a sleep.  It's so the test fails rather than hanging indefinitely if something goes horribly wrong.

dr7ana replied:
We don't do this in any other test, and any failure case where it hangs indefinitely should be handled internally?

jagerman replied:
> Are these sleeps

Not a sleep -- this just puts an upper limit on how long it can wait before returning with a non-ready status.

> We don't do this in any other test, and any failure case where it hangs indefinitely should be handled internally

We probably should -- if a failure happens in a test now the test suite hangs indefinitely, rather than failing, where failing would be nicer.

Perhaps putting this in a common test suite function that we can use in many places, something like:
```c++
template <typename F>
void require_future(F& f, std::chrono::milliseconds timeout = 1s) {
    REQUIRE(f.wait_for(timeout) == std::future_status::ready);
}
```
That said, 20ms seems way too short for this (because this is just a fail-safe in case it hangs).

dr7ana replied:
> This is a future wait with a timeout, rather than a sleep

> Not a sleep 

I was using "sleep" in the notional rather than literal sense

My main issue with this is using certificate verification as a proxy for "connection ready." We know already that this is not how ngtcp2 signals that at all

dr7ana commented on tests/008-ed-keys.cpp:
```diff
@@ -0,0 +1,147 @@
+        };
+
+        client_tls->set_key_verify_callback(client_key_allowed_cb);
+        server_tls2->set_key_verify_callback([&](auto, auto) {
```
Since you re-use this lambda 3 times, maybe you could define it where you defined `client_key_allowed_cb` and pass the lambda to `set_key_verify_callback`?

tewinget replied:
true, will do.

tewinget replied:
note that this only removes 1 duplicate though, as the other is returning false rather than true (and not based on input to the lambda)

jagerman commented on include/quic/gnutls_crypto.hpp:
```diff
@@ -23,6 +25,12 @@ namespace oxen::quic
             unsigned int incoming,
             const gnutls_datum_t* msg)>;
 
+    constexpr auto GNUTLS_KEY_SIZE = 32;  // for now, only supporting Ed25519 keys (32 bytes)
```
This might be better as an explicit `size_t` rather than auto (which makes it an `int`).

tewinget replied:
makes sense

jagerman commented on include/quic/gnutls_crypto.hpp:
```diff
@@ -69,7 +77,10 @@ namespace oxen::quic
+
+                // uint32_t cast to appease narrowing conversion gods,
+                // if cert size won't fit in 32 bits we have bigger problems
+                mem = {(uint8_t*)input.c_str(), (uint32_t)input.size()};
```
```suggestion
                mem = {reinterpret_cast<uint8_t*>(input.c_str()), static_cast<uint32_t>(input.size())};
```

tewinget replied:
so gnutls_datum_t is annoying and either requires a copy here into an allocated c-string or something *or* requires casting away const.  For simplicity, as it appears safe, this casts away const.  Could do that in the C++ way though,  I suppose.

tewinget committed:
genericize waiting on future with timeout in test suite

jagerman commented on tests/008-ed-keys.cpp:
```diff
@@ -0,0 +1,148 @@
+        };
+    };
+
+    TEST_CASE("008 - Ed Keys: Connections", "[001][keys][tls][connection]")
```
Perhaps name this "Handshake" rather than "Connections"?

tewinget replied:
agreed.

tewinget committed:
cpp-casts, clang-format, test name fixup

tewinget committed:
verify remote is correctly identified as client or relay

the lokinet-specific ALPN logic should be divorced from libquic, but for now
it is working; that can be changed later.

tewinget committed:
genericize ALPN handling

ALPN handling in raw public key mode is now handled generically,
rather than having Lokinet-specific concepts.  In future, it will be
easy to have ALPN handling for any mode, not just raw pubkey, but for
now that is not necessary as we do not use ALPNs with normal certs.
